gstreamer-video/src/auto/video_decoder.rs

```diff
@@ -0,0 +1,295 @@
+// This file was generated by gir (0fe730d) from gir-files (???)
+// DO NOT EDIT
+
+use VideoCodecFrame;
+use VideoCodecState;
+use VideoFormat;
+use ffi;
+use glib::object::IsA;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use gst;
+use gst_ffi;
+use std::mem;
+use std::ptr;
+
+glib_wrapper! {
+    pub struct VideoDecoder(Object<ffi::GstVideoDecoder>): [
+        gst::Element => gst_ffi::GstElement,
+        gst::Object => gst_ffi::GstObject,
+    ];
+
+    match fn {
+        get_type => || ffi::gst_video_decoder_get_type(),
+    }
+}
+
+unsafe impl Send for VideoDecoder {}
+unsafe impl Sync for VideoDecoder {}
+
+pub trait VideoDecoderExt {
+    fn add_to_frame(&self, n_bytes: i32);
```

Comment:
This, and various other functions must only be called while having taken the decoder stream lock

Reply:
So I take the lock myself in the imple, or how would you do that?

Reply:
I would have all methods that require the stream lock to be taken to do that inside the implementation. It's a recursive mutex, so doing it one time too often is not a problem apart from performance, and it would make this safe at least.

At a later time we could consider exposing unsafe functions that don't do that, for being able to implement a more performant solution. And also unsafe functions for unlocking the mutex (which is required if you want to block inside handle_frame()).